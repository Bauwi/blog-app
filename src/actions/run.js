import database from '../firebase/firebase';
import { db } from '../firebase/firebase';
import * as firebase from 'firebase';

// manage CRUD of run
export const addPostToRun = post => ({
  type: 'ADD_POST_TO_RUN',
  post
});

export const startAddPostToRun = post => (dispatch, getState) => {
  const { uid } = getState().auth;
  db
    .collection('runs')
    .doc(uid)
    .collection('UserRuns')
    .add(post)
    .then((doc) => {
      dispatch(addPostToRun({
        DBid: doc.id,
        ...post
      }));
    });
};

export const removePostToRun = id => ({
  type: 'REMOVE_POST_TO_RUN',
  id
});

export const startRemovePostToRun = (id, DBid) => (dispatch, getState) => {
  const { uid } = getState().auth;
  return db
    .collection('runs')
    .doc(uid)
    .collection('UserRuns')
    .doc(`${DBid}`)
    .delete()
    .then(dispatch(removePostToRun(id)));
};

// set Run posts. This action is dispatched in the header component
// as it needs to be available for the run dropdown

export const runsHasErrored = bool => ({
  type: 'RUN_HAS_ERRORED',
  hasErrored: bool
});

export const runIsLoading = bool => ({
  type: 'RUN_IS_LOADING',
  isLoading: bool
});

export const setRunPosts = posts => ({
  type: 'SET_RUN_POSTS',
  posts
});

export const startSetRunPosts = resume => (dispatch, getState) => {
  dispatch(runIsLoading(true));
  const { uid } = getState().auth;
  const run = [];
  return db
    .collection('runs')
    .doc(uid)
    .collection('UserRuns')
    .get()
    .then((querySnapshot) => {
      querySnapshot.forEach((doc) => {
        run.push({
          DBid: doc.id,
          ...doc.data()
        });
      });
      return run;
    })
    .then(run => dispatch(setRunPosts(run)))
    .then(() => {
      const firstUnreadIndex = run.findIndex(post => post.state === 'unread');
      if (firstUnreadIndex === -1) {
        return dispatch(setCurrentPostRun(run[0].content.id));
      }
      return dispatch(setCurrentPostRun(run[firstUnreadIndex].content.id));
    })
    .then(() => dispatch(runIsLoading(false)))
    .catch(() => dispatch(runsHasErrored(true)));
};

// change the status of the past from read to unread
export const updateRunPostToAlreadyRead = id => ({
  type: 'UPDATE_RUN_POST_TO_ALREADY_READ',
  id
});

export const startUpdateRunPostToAlreadyRead = (id, DBid) => (dispatch, getState) => {
  const { uid } = getState().auth;
  return db
    .collection('runs')
    .doc(uid)
    .collection('UserRuns')
    .doc(`${DBid}`)
    .update({ state: 'read' })
    .then(() => {
      dispatch(updateRunPostToAlreadyRead(id));
    });
};

// this one is not used anymore. But it could be used to mark some popst as unread yet
export const updateRunPostToNotAlreadyRead = id => ({
  type: 'UPDATE_RUN_POST_TO_NOT_ALREADY_READ',
  id
});

// Set which post should actually be current. Run navigation depends on this.
// instead of routing, component is updated with currentPost data.
// easier way to deal with "Resume" to provide the user the first unread post in its run.
export const setCurrentPostRun = id => ({
  type: 'SET_CURRENT_POST_RUN',
  id
});

export const startSetCurrentPostRun = id => (dispatch) => {
  dispatch(runIsLoading(true));
  return new Promise((resolve) => {
    setTimeout(() => {
      resolve();
    }, 10);
  })
    .then(() => dispatch(setCurrentPostRun(id)))
    .then(() => dispatch(runIsLoading(false)))
    .catch(() => dispatch(runsHasErrored(true)));
};

export const resetRun = () => ({
  type: 'RESET_RUN'
});

// remove all posts in tun
export const startResetRun = () => (dispatch, getState) => {
  const { uid } = getState().auth;
  return db
    .collection('runs')
    .doc(uid)
    .collection('UserRuns')
    .get()
    .then((querySnapshot) => {
      querySnapshot.forEach((doc) => {
        db
          .collection('runs')
          .doc(uid)
          .collection('UserRuns')
          .doc(doc.ref.id)
          .delete();
      });
    })
    .then(() => dispatch(resetRun()));
};

// Remove already read posts in run
export const cleanRun = () => ({
  type: 'CLEAN_RUN'
});

export const startCleanRun = alreadyRead => (dispatch, getState) => {
  const { uid } = getState().auth;
  return db
    .collection('runs')
    .doc(uid)
    .collection('UserRuns')
    .get()
    .then((querySnapShot) => {
      querySnapShot.forEach((doc) => {
        if (alreadyRead.includes(doc.ref.id)) {
          db
            .collection('runs')
            .doc(uid)
            .collection('UserRuns')
            .doc(doc.ref.id)
            .delete();
        }
      });
    })
    .then(dispatch(cleanRun()));
};
