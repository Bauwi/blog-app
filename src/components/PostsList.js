import React, { Component } from 'react';
import { connect } from 'react-redux';

import selectPosts from '../selectors/posts';

import PostListFilters from './PostListFilters';
import PostsListItem from './PostsListItem';

export class PostsList extends Component {
  renderPosts = () => {
    return this.props.posts.map(post => (
      <PostsListItem key={post.id} isAuthenticated={this.props.isAuthenticated} {...post} />
    ));
  };

  render() {
    return (
      <div>
        <PostListFilters />
        {this.renderPosts()}
      </div>
    );
  }
}

const mapStateToProps = state => ({
  posts: selectPosts(state.posts, state.filters),
  isAuthenticated: !!state.auth.uid
});

export default connect(mapStateToProps)(PostsList);
