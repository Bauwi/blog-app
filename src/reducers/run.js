const runReducerDefaultState = {
  isLoading: true, posts: [], hasErrored: false, current: {}
};

const runReducer = (state = runReducerDefaultState, action) => {
  switch (action.type) {
    case 'ADD_POST_TO_RUN':
      return {
        ...state,
        posts: [...state.posts, action.post]
      };
    case 'REMOVE_POST_TO_RUN':
      return {
        ...state,
        posts: state.posts.filter(post => action.id !== post.content.id)
      };
    case 'SET_RUN_POSTS':
      return {
        ...state,
        posts: action.posts
      };
    case 'RUN_IS_LOADING':
      return {
        ...state,
        isLoading: action.isLoading
      };
    case 'RUN_HAS_ERRORED':
      return {
        ...state,
        hasErrored: action.hasErrored
      };
    case 'SET_CURRENT_POST_RUN':
      return {
        ...state,
        current: state.posts.find(post => post.content.id === action.id)
      };
    case 'UPDATE_RUN_POST_TO_ALREADY_READ':
      return {
        ...state,
        current: { ...state.current, state: 'read' },
        posts: state.posts.map((post) => {
          if (post.content.id === action.id) {
            return {
              ...post,
              state: 'read'
            };
          }
          return post;
        })
      };
    case 'UPDATE_RUN_POST_TO_NOT_ALREADY_READ':
      return {
        ...state,
        posts: state.posts.map((post) => {
          if (post.content.id === action.id) {
            return {
              ...post,
              state: 'unread'
            };
          }
          return post;
        })
      };
    case 'RESET_RUN':
      return {
        posts: []
      };
    case 'CLEAN_RUN':
      return {
        ...state,
        posts: state.posts.filter(post => post.state === 'unread')
      };
    default:
      return state;
  }
};

export default runReducer;
