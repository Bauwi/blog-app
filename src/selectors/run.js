const getRunPosts = (posts = [], run) => run.map(read => posts.find(post => post.id === read.id));
export default getRunPosts;
