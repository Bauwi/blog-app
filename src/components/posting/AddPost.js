import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import PostForm from './PostForm';
import { startAddPost } from '../../actions/posts';
import { startUpdateUser } from '../../actions/users';

export class AddPost extends Component {
  onSubmit = post => {
    this.props.startAddPost(post);
    this.props.startUpdateUser(undefined, { username: post.author });
    this.props.history.push('/');
  };

  render() {
    return (
      <div>
        <div className="page-header">
          <header className="content-container">
            <Link to="/dashboard">Back to dashboard</Link>
          </header>
        </div>
        <PostForm onSubmit={this.onSubmit} context="add" author={this.props.user} />
      </div>
    );
  }
}

const mapDispatchToProps = dispatch => ({
  startAddPost: post => dispatch(startAddPost(post)),
  startUpdateUser: (id, updates) => dispatch(startUpdateUser(id, updates))
});

const mapStateToProps = state => ({
  user: state.users.preferences
});
export default connect(mapStateToProps, mapDispatchToProps)(AddPost);
