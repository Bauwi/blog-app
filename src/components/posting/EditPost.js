import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { startEditPost, startRemovePost } from '../../actions/posts';
import { startUpdateUser } from '../../actions/users';

import PostForm from './PostForm';

export class EditPost extends Component {
  onSubmit = post => {
    this.props.startEditPost(this.props.post.id, post);
    this.props.history.push('/dashboard');
  };
  onRemovePost = () => {
    this.props.startRemovePost(this.props.post.id);
    this.props.startUpdateUser(undefined, {
      numberOfPosts: this.props.preferences.numberOfPosts - 1
    });
    this.props.history.push('/dashboard');
  };

  render() {
    return (
      <div>
        <PostForm
          onSubmit={this.onSubmit}
          post={this.props.post}
          author={this.props.preferences}
          context="edit"
        />
        <div className="content-container">
          <button className="button" onClick={this.onRemovePost}>
            Remove Post
          </button>
        </div>
      </div>
    );
  }
}

EditPost.propTypes = {
  post: PropTypes.shape({
    id: PropTypes.string.isRequired,
    author: PropTypes.string.isRequired,
    authorId: PropTypes.string.isRequired,
    body: PropTypes.arrayOf(PropTypes.object).isRequired,
    cover: PropTypes.string.isRequired,
    createdAt: PropTypes.number.isRequired,
    keywords: PropTypes.string.isRequired,
    miniCover: PropTypes.string.isRequired,
    readingTime: PropTypes.number.isRequired,
    title: PropTypes.string.isRequired
  }).isRequired,
  preferences: PropTypes.shape({
    avatar: PropTypes.string,
    description: PropTypes.string,
    email: PropTypes.string,
    stars: PropTypes.number.isRequired,
    topCategories: PropTypes.arrayOf(PropTypes.string),
    username: PropTypes.string
  }).isRequired,
  startEditPost: PropTypes.func.isRequired,
  startRemovePost: PropTypes.func.isRequired
};

const mapStateToProps = (state, props) => {
  return {
    post: state.posts.posts.find(post => {
      return post.id === props.match.params.id;
    }),
    preferences: state.users.preferences
  };
};

const mapDispatchToProps = (dispatch, props) => ({
  startEditPost: (id, post) => dispatch(startEditPost(id, post)),
  startRemovePost: id => dispatch(startRemovePost(id)),
  startUpdateUser: (id, updates) => dispatch(startUpdateUser(id, updates))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditPost);
