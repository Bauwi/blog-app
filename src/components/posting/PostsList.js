import React, { Component } from 'react';
import { connect } from 'react-redux';

import { startSetPosts } from '../../actions/posts';
import selectPosts from '../../selectors/posts';

import PostListFilters from '../PostListFilters';
import PostsListItem from './PostsListItem';

export class PostsList extends Component {
  state = {
    loaded: 10
  };
  handleMorePosts = () => {
    this.props.startSetPosts(undefined, this.state.loaded + 10);
    this.setState(prevState => ({ loaded: prevState + 12 }));
  };

  renderPosts = () => {
    return this.props.posts
      .sort((a, b) => a.createdAt - b.createdAt)
      .reverse()
      .map(post => <PostsListItem key={post.id} {...post} />);
  };

  render() {
    if (this.props.posts.length === 0) {
      return (
        <div className="content-container">
          <PostListFilters />
          <p>No posts found here</p>
        </div>
      );
    }
    return (
      <div>
        <PostListFilters />
        <div className="grid-dashboard">{this.renderPosts()}</div>
        <button className="button button--more" onClick={this.handleMorePosts}>
          Load more
        </button>
      </div>
    );
  }
}

const mapDispatchToProps = dispatch => ({
  startSetPosts: (id, rangeMax) => dispatch(startSetPosts(id, rangeMax))
});

const mapStateToProps = state => ({
  posts: selectPosts(state.posts, state.filters)
});

export default connect(mapStateToProps, mapDispatchToProps)(PostsList);
