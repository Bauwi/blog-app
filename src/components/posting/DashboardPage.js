import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import PostsList from './PostsList';
import PostsSummary from './PostsSummary';
import UserCard from '../UserCard';

const DashboardPage = ({ preferences, userId }) => (
  <div>
    <PostsSummary />

    <div className="content-container">
      <UserCard author={preferences} userId={userId} />
      <PostsList />
    </div>
  </div>
);

DashboardPage.propTypes = {
  preferences: PropTypes.shape({
    avatar: PropTypes.string,
    description: PropTypes.string,
    email: PropTypes.string,
    stars: PropTypes.number.isRequired,
    topCategories: PropTypes.arrayOf(PropTypes.string),
    username: PropTypes.string
  }).isRequired,
  userId: PropTypes.string.isRequired
};

const mapStateToProps = state => ({
  preferences: state.users.preferences,
  userId: state.auth.uid
});

export { DashboardPage };
export default connect(mapStateToProps)(DashboardPage);
