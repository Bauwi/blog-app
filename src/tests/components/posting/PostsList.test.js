import React from 'react';
import { shallow, mount } from 'enzyme';
import { PostsList } from '../../../components/posting/PostsList';
import run from '../../fixtures/run';

const posts = run.posts.map(post => post.content);

test('should render PostsLists component correctly on loading', () => {
  // const startSetPosts = jest.fn();
  // const sortByDate = jest.fn()
  const wrapper = shallow(<PostsList isLoading posts={posts} startSetPosts={() => {}} sortByDate={() => {}} />);
  expect(wrapper).toMatchSnapshot();
});

test('should render PostsLists component correctly after loading', () => {
  const wrapper = shallow(<PostsList isLoading={false} posts={posts} startSetPosts={() => {}} sortByDate={() => {}} />);
  expect(wrapper).toMatchSnapshot();
});

test('should call sortByDate on mount', () => {
  const sortByDate = jest.fn();
  const wrapper = shallow(<PostsList isLoading={false} posts={posts} startSetPosts={() => {}} sortByDate={sortByDate} />);
  expect(sortByDate).toHaveBeenCalled();
});

// test('should call startSetPosts on mount', () => {
//   const startSetPosts = jest.fn();
//   const wrapper = mount(<PostsList posts={posts} startSetPosts={startSetPosts} sortByDate={() => {}} />);
//   ReactTestUtils.Simulate.scroll(wrapper, { deltaY: 500 });
//   wrapper.simulate('scroll');
//   expect(startSetPosts).toHaveBeenCalled();
// });
