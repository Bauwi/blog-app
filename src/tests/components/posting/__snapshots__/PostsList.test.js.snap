// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render PostsLists component correctly after loading 1`] = `
<div
  className="content-container post-list"
>
  <PostListFilters />
  <div
    className="grid-dashboard"
  >
    <PostsListItem
      author="author1"
      authorId="author1Id"
      body={
        Array [
          Object {
            "attributes": "h1",
            "text": "This is the first Post in run",
          },
        ]
      }
      cover="https://www.w3schools.com/w3css/img_fjords.jpg"
      createdAt={0}
      id="1"
      key="1"
      keywords="red, green, blue"
      miniCover="https://www.w3schools.com/w3css/img_fjords.jpg"
      readingTime={7}
      stars={5}
      title="Post 1 in Run"
    />
    <PostsListItem
      author="author2"
      authorId="author2Id"
      body={
        Array [
          Object {
            "attributes": "h2",
            "text": "This is the second Post in run",
          },
        ]
      }
      cover="https://www.w3schools.com/w3css/img_fjords.jpg"
      createdAt={-1500}
      id="2"
      key="2"
      keywords="cat, dog, fish"
      miniCover="https://www.w3schools.com/w3css/img_fjords.jpg"
      readingTime={12}
      stars={25}
      title="Post 2 in Run"
    />
    <PostsListItem
      author="author3"
      authorId="author3Id"
      body={
        Array [
          Object {
            "attributes": "h3",
            "text": "blabla",
          },
        ]
      }
      cover="https://www.w3schools.com/w3css/img_fjords.jpg"
      createdAt={1000000}
      id="3"
      key="3"
      keywords="car, pool, cake"
      miniCover="https://www.w3schools.com/w3css/img_fjords.jpg"
      readingTime={2}
      stars={2}
      title="Post 3 in Run"
    />
  </div>
  <div
    className="loader loader--small loader--empty"
  />
</div>
`;

exports[`should render PostsLists component correctly on loading 1`] = `
<div
  className="content-container post-list"
>
  <PostListFilters />
  <div
    className="grid-dashboard"
  >
    <PostsListItem
      author="author1"
      authorId="author1Id"
      body={
        Array [
          Object {
            "attributes": "h1",
            "text": "This is the first Post in run",
          },
        ]
      }
      cover="https://www.w3schools.com/w3css/img_fjords.jpg"
      createdAt={0}
      id="1"
      key="1"
      keywords="red, green, blue"
      miniCover="https://www.w3schools.com/w3css/img_fjords.jpg"
      readingTime={7}
      stars={5}
      title="Post 1 in Run"
    />
    <PostsListItem
      author="author2"
      authorId="author2Id"
      body={
        Array [
          Object {
            "attributes": "h2",
            "text": "This is the second Post in run",
          },
        ]
      }
      cover="https://www.w3schools.com/w3css/img_fjords.jpg"
      createdAt={-1500}
      id="2"
      key="2"
      keywords="cat, dog, fish"
      miniCover="https://www.w3schools.com/w3css/img_fjords.jpg"
      readingTime={12}
      stars={25}
      title="Post 2 in Run"
    />
    <PostsListItem
      author="author3"
      authorId="author3Id"
      body={
        Array [
          Object {
            "attributes": "h3",
            "text": "blabla",
          },
        ]
      }
      cover="https://www.w3schools.com/w3css/img_fjords.jpg"
      createdAt={1000000}
      id="3"
      key="3"
      keywords="car, pool, cake"
      miniCover="https://www.w3schools.com/w3css/img_fjords.jpg"
      readingTime={2}
      stars={2}
      title="Post 3 in Run"
    />
  </div>
  <Component />
</div>
`;
